src/fuzzing.rs

```diff
@@ -0,0 +1,227 @@
+use std::sync::Arc;
+
+use arbitrary::*;
+
+use crate::prelude::*;
+
+// Adding new fuzz targets
+//
+// 1. Add a function to this file which looks like this:
+//
+//    pub fn fuzz_something(data: &[u8], debug: bool) {
+//      let mut g = create_generator!();
+//
+//      // Invoke everything you need.
+//      //
+//      // You should use g.g() to get an arbitrary value of any type that
+//      // implements Arbitrary [1]. This is how fuzzer affects the
+//      // execution—by feeding in different bytes, which result in different
+//      // arbitrary values being generated.
+//      // [1]: https://docs.rs/arbitrary/0.2.0/arbitrary/trait.Arbitrary.html
+//      //
+//      // Print out the structures you create with arbitrary data when debug
+//      // is set to true. This is for running individual artifacts with
+//      // FUZZ_DEBUG.
+//    }
+//
+// 2. cargo fuzz add something
+// 3. Copy the contents of any other .rs file from fuzz/fuzz_targets/ into the
+//    newly created fuzz/fuzz_targets/something.rs and change the function
+//    being called to fuzz_something.
+//
+// Now you can fuzz the new target with cargo fuzz.
+
+// A helper for generating arbitrary data.
+struct Generator<'a> {
+  buffer: RingBuffer<'a>,
+}
+
+impl<'a> Generator<'a> {
+  fn new(data: &'a [u8]) -> Result<Self, BufferError> {
+    Ok(Self { buffer: RingBuffer::new(data, data.len())? })
+  }
+
+  fn g<T: Arbitrary>(&mut self) -> T {
+    <T as Arbitrary>::arbitrary(&mut self.buffer).unwrap()
+  }
+}
+
+macro_rules! create_generator {
+  ($data:expr) => {{
+    let g = Generator::new($data);
+    if g.is_err() {
+      return;
+    }
+    g.unwrap()
+  }};
+}
+
+pub fn fuzz_construct_context(data: &[u8], debug: bool) {
+  let mut g = create_generator!(data);
+
+  let mut config = Config::default();
+  config.threads = 1;
+  config.enc.width = g.g();
+  config.enc.height = g.g();
+  config.enc.bit_depth = (g.g::<u8>() % 17) as usize;
+  config.enc.still_picture = g.g();
+  config.enc.time_base = Rational::new(g.g(), g.g());
+  config.enc.min_key_frame_interval = g.g();
+  config.enc.max_key_frame_interval = g.g();
+  config.enc.reservoir_frame_delay = g.g();
+  config.enc.low_latency = g.g();
+  config.enc.quantizer = g.g();
+  config.enc.min_quantizer = g.g();
+  config.enc.bitrate = g.g();
+  config.enc.tile_cols = g.g();
+  config.enc.tile_rows = g.g();
+  config.enc.tiles = g.g();
+  config.enc.rdo_lookahead_frames = g.g();
+  config.enc.speed_settings = SpeedSettings::from_preset(g.g());
+  config.enc.show_psnr = g.g();
+  config.enc.train_rdo = g.g();
+
+  if debug {
```

Comment:
Instea of having these `if debug` blocks, would it possible to wrap this in a macro or logging function that is a noop if the debug flag is disabled?

Reply:
^ Now that we have the `log` crate, you can use the `debug!` macro instead of `println!`.

Reply:
^ Now that we have the log crate in rav1e, you can use the `debug!` macro in place of `println!` to do this.